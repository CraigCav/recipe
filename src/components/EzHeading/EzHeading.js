import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';
import {size, Subheading} from './EzHeading.styles';
import {standard} from '../../themes';

/**
 * Headings are used to create visual hierarchy in page content. They are the primary means of controlling typography.
 */
const EzHeading = ({as, children, className, size: headingSize, subheading, theme}) => {
  const headingElement = as || 'h' + headingSize;
  const Heading = styled(headingElement)(size);
  const useSubheading = subheading && (headingSize === '3' || headingSize === '5');

  return (
    <Heading className={className} size={headingSize} theme={theme}>
      {children}
      {useSubheading && <Subheading theme={theme}>{subheading}</Subheading>}
    </Heading>
  );
};

EzHeading.propTypes = {
  /**
   * An optional property allowing you to override which tag is rendered. Use sparingly for accessiblity or SEM purposes.
   */
  as: PropTypes.string,
  /**
   * The text for the heading.
   */
  children: PropTypes.string.isRequired,
  /**
   * The size of heading to render.
   */
  size: PropTypes.oneOf(['1', '2', '3', '4', '5', '6']).isRequired,
  /**
   * An optional subheading. Only supported with size 3 and size 5 headings.
   */
  subheading: PropTypes.string,
};

EzHeading.defaultProps = {
  theme: standard,
};

/**
 * @component
 */
export default styled(EzHeading)();
