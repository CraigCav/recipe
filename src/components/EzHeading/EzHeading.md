---
name: EzHeading
title: Heading
path: "/components/ez-heading"
---

Headings are the primary way of styling text in Recipe. Use them to create visual hierarchy on the page.

---

## Best Practices

Headings should:

* Be used to style text.
* Follow Recipe's guidance on [typography](/styles/style#typography).
* Use the `as` property if they need to provide a hierarchy for accessibility or SEM reasons that is different from the visual hierarchy.

Headings should not:

* Use the `as` property unless there there is a specific need to have a visual hierarchy that is different from the markup hierarchy.

---

## Content guidelines

Use sentence case for content in headings (only capitalize the first word).

## Examples

### Basic Ez heading

We support six sizes of heading. Headings don't have margin...in our current model margin is applied by containers to their children.

```jsx live
<React.Fragment>
  <EzHeading size="1">Page titles</EzHeading>
  <EzHeading size="2">On-page headings</EzHeading>
  <EzHeading size="3">Container headings</EzHeading>
  <EzHeading size="4">Navigation / tabs</EzHeading>
  <EzHeading size="5">Labels</EzHeading>
  <EzHeading size="6">Small stuff</EzHeading>
</React.Fragment>
```

You can provide an optional subheading. At this time you should only use subheadings with `3` and `5` headings. If you provide a `subheading` prop to other sizes it will not be rendered.

```jsx live
<EzCard>
  <EzHeading size="3" subheading="I'm a subheading">Container headings</EzHeading>
  <EzHeading size="5" subheading="I'm a subheading">Labels</EzHeading>
</EzCard>
```

Normally heading size numbers correspond directly to the same size `h` tag, but you can provide an optional `as` property to override the markup that is rendered for heading. You should only need to use this property sparingly and only for accessibility or SEM purposes. For example: an empty state might call for a visually larger heading, but it's position in the page hierarchy might make a lower level `h` tag the appropriate markup.

```jsx live
<EzHeading size="3" as="h1">Container headings</EzHeading>
```

---

## Related components

\*
