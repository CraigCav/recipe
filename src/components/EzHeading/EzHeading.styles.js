import styled, {css} from 'react-emotion';
import variant from 'styled-component-variant';

export const Subheading = styled.div`
  color: ${props => props.theme.colors.grays[600]};
  font-size: ${props => props.theme.fontSizes[300]};
  font-weight: ${props => props.theme.fontWeights.normal};
`;

const heading = (size, weight) => ({theme}) => css`
  font-size: ${theme.fontSizes[size]};
  font-weight: ${theme.fontWeights[weight]};
  margin: 0;
`;

const size = variant('size', {
  1: heading(700, 'normal'),
  2: heading(600, 'normal'),
  3: heading(500, 'bold'),
  4: heading(400, 'normal'),
  5: heading(300, 'bold'),
  6: heading(200, 'normal'),
});

export const headings = ['1', '2', '3', '4', '5', '6'].reduce((res, n) => {
  const tag = `h${n}`;
  return Object.assign({[tag]: styled(tag)(size)}, res);
}, {});
