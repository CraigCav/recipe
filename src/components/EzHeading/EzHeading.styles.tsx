import variant from 'styled-component-variant';
import {css} from '@emotion/core';
import {align} from '../EzTextStyle/EzTextStyle.styles';
import styled from '../../themes/styled';

export const Subheading = styled.div`
  color: ${props => props.theme.colors.text.deemphasis};
  font-size: ${props => props.theme.fontSizes[300]};
  font-weight: ${props => props.theme.fontWeights.normal};
  line-height: ${props => props.theme.lineHeights.heading};
  margin-top: ${props => props.theme.spacing.xs};
`;

const heading = (size, weight) => ({theme}) => css`
  color: ${theme.colors.text.base};
  font-size: ${theme.fontSizes[size]};
  font-weight: ${theme.fontWeights[weight]};
  line-height: ${theme.lineHeights.heading};
  margin: 0;
`;

const size = variant('size', {
  1: heading(700, 'normal'),
  2: heading(600, 'normal'),
  3: heading(500, 'bold'),
  4: heading(400, 'normal'),
  5: heading(300, 'bold'),
  6: heading(200, 'normal'),
});

export const headings = ['1', '2', '3', '4', '5', '6'].reduce((res, n) => {
  const tag = `h${n}` as 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
  return {[tag]: styled(tag)(size, align), ...res};
}, {});

export const Header = styled.header(align);
