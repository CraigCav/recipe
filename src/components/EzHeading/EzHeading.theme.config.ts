import {mergeCss} from '../theme.config';

export default mergeCss({
  theme: {
    colors: {
      'heading-text': '$text',
      'subheading-text': '$deemphasisText',
    },
    fontSizes: {
      subheading: '$100',
    },
    fontWeights: {
      subheading: '$400',
    },
    lineHeights: {
      heading: '$snug',
      subheading: '$snug',
    },
  },
});
