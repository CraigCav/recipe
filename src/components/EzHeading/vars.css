:root {
  --recipe-heading-text-color: var(--recipe-alias-text-color);
  --recipe-heading-leading: var(--recipe-alias-leading-title);
  --recipe-subheading-text-color: var(--recipe-alias-deemphasis-text-color);
  --recipe-subheading-font-size: var(--recipe-global-font-size-100);
  --recipe-subheading-font-weight: var(--recipe-global-font-weight-regular);
  --recipe-subheading-leading: var(--recipe-alias-leading-title);
}