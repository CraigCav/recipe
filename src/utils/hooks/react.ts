import React from 'react';

let currentInstance = null;

export const use = <P>(render: React.SFC<P>): React.ComponentClass<P> => {
  return class extends React.Component<P> {
    state = {};
    effects = [];
    componentDidMount() {
      this.afterRender();
    }
    componentDidUpdate() {
      this.afterRender();
    }
    componentWillUnmount() {
      this.effects.forEach(({destroy}) => {
        if (destroy) destroy();
      });
    }
    afterRender() {
      this.effects.forEach(({create, destroy}) => {
        if (destroy) destroy();
        create();
      });
    }
    render() {
      currentInstance = this;
      Object.assign(currentInstance, {callIndex: 0, effects: []});
      return render(this.props);
    }
  };
};

const nextStateCell = () => {
  const frame = currentInstance;
  const {state, callIndex} = currentInstance;
  currentInstance.callIndex++;
  return {
    frame,
    state,
    key: callIndex,
  };
};

export const useRef = <TValue>(initialValue: TValue): {current: TValue} => {
  const {state, key} = nextStateCell();

  if (!state.hasOwnProperty(key)) {
    state[key] = {current: initialValue};
  }

  return state[key];
};

export const useState = <TValue>(initialValue: TValue): [TValue, (value: TValue) => void] => {
  const {frame, state, key} = nextStateCell();

  return [
    state.hasOwnProperty(key) ? state[key] : initialValue,
    newValue => frame.setState({[key]: newValue}),
  ];
};

const inputsAreEqual = (next, prev) => next && next.every((d, i) => d === prev[i]);

export const useEffect = (create, inputs = [create]) => {
  const {frame, state, key} = nextStateCell();

  const prevEffect = state[key];

  if (prevEffect && inputsAreEqual(inputs, prevEffect.inputs)) return;

  const effect = (state[key] = {
    create,
    inputs,
    destroy: prevEffect ? prevEffect.destroy : null,
  });

  frame.effects.push(effect);
};
