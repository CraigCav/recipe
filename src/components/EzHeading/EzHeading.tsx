import React from 'react';
import PropTypes from 'prop-types';
import {headings, Subheading} from './EzHeading.styles';
import styled from '../../themes/styled';

/**
 * Headings are used to create visual hierarchy in page content. They are the primary means of controlling typography.
 */
const EzHeading = ({as, children, className, size: headingSize, subheading}) => {
  const headingElement = as || `h${headingSize}`;
  const Heading = headings[headingElement];
  const useSubheading = subheading && (headingSize === '3' || headingSize === '5'); //

  return (
    <Heading className={className} size={headingSize}>
      {children}
      {useSubheading && <Subheading>{subheading}</Subheading>}
    </Heading>
  );
};

EzHeading.propTypes = {
  /**
   * An optional property allowing you to override which tag is rendered. Use sparingly for accessiblity or SEM purposes.
   */
  as: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  /**
   * The text for the heading.
   */
  children: PropTypes.string.isRequired,
  /**
   * The size of heading to render.
   */
  size: PropTypes.oneOf(['1', '2', '3', '4', '5', '6']).isRequired,
  /**
   * An optional subheading. Only supported with size 3 and size 5 headings.
   */
  subheading: PropTypes.string,
};

/**
 * @component
 */
//@ts-ignore
export default styled(EzHeading)();
