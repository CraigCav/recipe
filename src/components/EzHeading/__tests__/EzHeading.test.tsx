import React from 'react';
import {axe} from 'jest-axe';
import {visualSnapshots} from 'sosia';
import {render} from '@testing-library/react';
import markdown from '../EzHeading.md';
import EzHeading from '../EzHeading';
import {EzCard} from '../../index';

const scope = {EzHeading, EzCard, React};

describe('EzHeading', () => {
  visualSnapshots({markdown, scope});

  it('should render with the specified tag', () => {
    const {container} = render(
      <EzHeading as="h1" size="6">
        Heading
      </EzHeading>
    );
    expect(container.querySelectorAll('h1')).toHaveLength(1);
  });

  test.each(['1', '2', '3', '4', '5', '6'])('Shows/Hides subheading for size %i', n => {
    const size = n as '1' | '2' | '3' | '4' | '5' | '6';

    const {container} = render(
      <EzHeading size={size} subheading="Subheading">
        Heading
      </EzHeading>
    );
    const shown = n === '3' || n === '5';

    if (shown) expect(container).toHaveTextContent(/Subheading/);
    else expect(container).not.toHaveTextContent(/Subheading/);
  });

  it('should meet accessibility guidelines', async () => {
    const {container} = render(<EzHeading size="1">Heading</EzHeading>);
    const actual = await axe(container.outerHTML);
    expect(actual).toHaveNoViolations();
  });
});
