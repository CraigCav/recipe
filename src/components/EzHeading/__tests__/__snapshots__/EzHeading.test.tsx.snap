// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EzHeading should render size 1 with size 700, weight normal 1`] = `
.recipe-1 {
  font-size: 2.2857142857142856rem;
  font-weight: 400;
  margin: 0;
}

<h1
  className="recipe-0 recipe-1 recipe-2"
  size="1"
>
  Heading
</h1>
`;

exports[`EzHeading should render size 2 with size 600, weight normal 1`] = `
.recipe-1 {
  font-size: 1.5rem;
  font-weight: 400;
  margin: 0;
}

<h2
  className="recipe-0 recipe-1 recipe-2"
  size="2"
>
  Heading
</h2>
`;

exports[`EzHeading should render size 3 with size 500, weight bold 1`] = `
.recipe-1 {
  font-size: 1.2857142857142858rem;
  font-weight: 700;
  margin: 0;
}

<h3
  className="recipe-0 recipe-1 recipe-2"
  size="3"
>
  Heading
</h3>
`;

exports[`EzHeading should render size 4 with size 400, weight normal 1`] = `
.recipe-1 {
  font-size: 1.1428571428571428rem;
  font-weight: 400;
  margin: 0;
}

<h4
  className="recipe-0 recipe-1 recipe-2"
  size="4"
>
  Heading
</h4>
`;

exports[`EzHeading should render size 5 with size 300, weight bold 1`] = `
.recipe-1 {
  font-size: 1rem;
  font-weight: 700;
  margin: 0;
}

<h5
  className="recipe-0 recipe-1 recipe-2"
  size="5"
>
  Heading
</h5>
`;

exports[`EzHeading should render size 6 with size 200, weight normal 1`] = `
.recipe-1 {
  font-size: 0.8571428571428571rem;
  font-weight: 400;
  margin: 0;
}

<h6
  className="recipe-0 recipe-1 recipe-2"
  size="6"
>
  Heading
</h6>
`;
